File: kelvinWu/quiz/q1.js
// Given an array of positive numbers and a positive number ‘k,’ find the maximum sum of any contiguous subarray of size ‘k’.

// [1, 2, 3, 4, 5] k = 3

// 1 2 3    2 3 4    3 4 5   need to check each of these subarrays

// each subarray only change one number at a time

// can use sliding window to travese through Array

//keep a running sum of subarray currently looking at
//keep maxSum variable to keep track of best subarray sum found so far
//loop through array
////if index < k, add it to sum
////else if we are >= k i.e. we have reached the k window size for the first k elements already, need to slide the window
//////delete the number k away from currentIndex
//////add currentNumber to index
////after add adding and/or subtracting numbers, check if current subarraysum is greater than maxSum.
//////replace maxSum with currentSum

// k = 2
// i:   0 1 2 3 4 5
// arr: 1 2 3 4 5 6

function maxSubarraySum(arr, k) {
  let maxSum = 0;
  let currentSum = 0;

  for (let i = 0; i < arr.length; i++) {
    if (i >= k) {
      let indexToSubtract = i - k;
      currentSum -= arr[indexToSubtract];
    }
    currentSum += arr[i];
    maxSum = Math.max(maxSum, currentSum);
  }

  return maxSum;
}

console.log(
  `[${[1, 2, 3, 4, 5, 6]}]  k: ${1}  answer: ${maxSubarraySum(
    [1, 2, 3, 4, 5, 6],
    1
  )}`
);

console.log(
  `[${[1, 2, 6, 4, 5, 3]}]  k: ${2}  answer: ${maxSubarraySum(
    [1, 2, 6, 4, 5, 3],
    2
  )}`
);

console.log(
  `[${[1, 1, 1, 1, 1, 1]}]  k: ${3}  answer: ${maxSubarraySum(
    [1, 1, 1, 1, 1, 1],
    3
  )}`
);

console.log(
  `[${[1, 1, 1, 1, 1, 1]}]  k: ${6}  answer: ${maxSubarraySum(
    [1, 1, 1, 1, 1, 1],
    6
  )}`
);

File: mariannaFervenza/Quiz/Q1.py
# Given an array of positive numbers and a positive number ‘k,’ find the maximum sum of any contiguous subarray of #size ‘k’.


# So we are given an array of k = [positive numbers]
# and we are given a variable k = "integer that is the size of the array"
# We need to find the max sum of of this subarrays with size k

def maxSub(arr, k):
    length = len(arr)
    index = 0

    if length == 0:
        return 0

    result = 0
    maxSum = result
    while index >= k:

        maxSum += arr[index] - arr[index - k]
        result = max(result, maxSum)

        return result

    if length < k:
        return "Invalid code"


def main():
    arr = [2, 4, 6, 7, 8, 10, 20]
    k = 4
    print(maxSub(arr, k))
    arr1 = []
    k1 = 1
    print(maxSub(arr1, k1))
    arr2 = [5, 10, 15, 25]
    k2 = 3
    print(maxSub(arr2, k2))


main()
